powersimdata/input/converter/pypsa_to_grid.py

```diff
@@ -286,13 +282,77 @@ def _read_network(self, n, add_pypsa_cols=True):
-        storage_storagedata_storageunits = _get_storage_storagedata(n, "storage_units")
-        storage_gen_stores = _get_storage_gen(n, "stores")
-        storage_gencost_stores = _get_storage_gencost(n, "stores")
-        storage_storagedata_stores = _get_storage_storagedata(n, "stores")
+        # storage units
+        c = "storage_units"
+        storage_gen_storageunits = _get_storage_gen(n.storage_units, c)
+        storage_gencost_storageunits = _get_storage_gencost(n.storage_units, c)
+        storage_storagedata_storageunits = _get_storage_storagedata(n.storage_units, c)
+
+        inflow = n.get_switchable_as_dense("StorageUnit", "inflow")
+        has_inflow = inflow.any()
+        if has_inflow.any():
+            # add artificial buses
+            suffix = " inflow"
+
+            def add_suffix(s):
+                return str(s) + suffix
+
+            storage_gen_inflow = storage_gen_storageunits[has_inflow]
+            buses_old = storage_gen_inflow.bus_id
+            buses_new = buses_old.astype(str) + suffix
+            bus_inflow = bus.reindex(buses_old).rename(index=add_suffix)
+
+            # add discharging dcline (has same index as inflow storages)
+            dcline_inflow = pd.DataFrame(
+                {
+                    "from_bus_id": buses_new,
+                    "to_bus_id": buses_old,
+                    "Pmax": storage_gen_inflow.Pmax,
+                }
+            )
+            dcline_inflow.index += " discharge"
+
+            # add charging dcline (has same index as inflow storages)
+            dcline_inflow = pd.DataFrame(
+                {
+                    "from_bus_id": buses_old,
+                    "to_bus_id": buses_new,
+                    "Pmax": -storage_gen_inflow.Pmin,
+                }
+            )
+            dcline_inflow.index += " charge"
+
+            # add inflow generator
+            gen_inflow = storage_gen_inflow.rename(index=add_suffix)
+            gen_inflow["Pmax"] = n.storage_units_t.inflow.max().rename(add_suffix)
+            gen_inflow["capital_cost"] = 0
+            gen_inflow["p_nom_extendable"] = False
+            gen_inflow = gen_inflow.reindex(columns=plant.columns)
+            gencost_inflow = storage_gencost_storageunits[has_inflow].rename(
+                index=add_suffix
+            )
+            gencost_inflow = storage_gencost_storageunits[has_inflow].rename(
+                index=add_suffix
+            )
+            gencost_inflow = storage_gencost_storageunits.assign(
+                c0=0, c1=0, c2=0, type=2, startup=0, shutdown=0, n=3
+            )
+
+            # add everything to data
+            storage_gen_storageunits.loc[has_inflow, "bus_id"] = buses_new
+            storage_gen_storageunits.loc[
+                has_inflow, "Pmin"
+            ] = -np.inf  # don't limit charging from inflow
+            bus = pd.concat([bus, bus_inflow])
+            plant = pd.concat([plant, gen_inflow])
```

Comment:
This raises the following warning:
```
FutureWarning: Behavior when concatenating bool-dtype and numeric-dtype arrays is deprecated; in a future version these will cast to object dtype (instead of coercing bools to numeric values). To retain the old behavior, explicitly cast bool-dtype arrays to numeric dtype.
  plant = pd.concat([plant, gen_inflow])
```

Reply:
Is fixed now. "Unresovle" if there is still something missing.